### PR TITLE
fixed ARI (fixes #225 & #226)

### LINKED ISSUE #225: Adjusted Rand index inconsistency for large n
The adjusted Rand index fails unexpectedly when n is large (n > 100,000). Here is an example with a comparison to an R implementation.
```
using Random
using Clustering
using RCall
Random.seed!(123);

n = 100_000;
a = rand(1:3,n);
b = rand(1:3,n);

randindex(a,b)[1]

only(R"library(mclust); adjustedRandIndex($a,$b)")
```
which gives
```
0.2933142400616828

-1.5731751561282826e-6
```
In theory the true adjusted Rand index should be close to 0. This starts to happen around `n=83,000` for me.

As a Julia comparison, my own implementation of the adjusted Rand index gives the same result as in R:
```
function ari(a,b)
    table = counts(a,b)
    acounts = sum(table,dims=1)
    bcounts = sum(table,dims=2)
    
    score = sum([x*(x-1)/2 for x in table])
    asum = sum([x*(x-1)/2 for x in acounts])
    bsum = sum([x*(x-1)/2 for x in bcounts])
    expected = asum*bsum/binomial(sum(table),2)
    total = (asum + bsum)/2
    
    if total == expected
        return 0
    else
        return (score-expected)/(total-expected)
    end
end;
ari(a,b)
```
```
-1.5731751561282826e-6
```
I use Clustering.jl 0.14.2, Julia 1.6.2.

### PR DESCRIPTION
**Pull request by wildart**
- added pair confusion matrix (CM) calculations
- fixed ARI calculation using CM

### CODE REVIEW

**Comment by wildart** [comment]
As `rand` behavior changed in recent julia versions, the rework of RNG handling or a test review is required.

**Comment by pcjentsch** [comment]
This PR passes the relevant tests, can it be merged?

This is an old bug that has had other PRs to fix it: https://github.com/JuliaStats/Clustering.jl/pull/99

**Comment by ararslan** [comment]
Looks like tests are passing except for 32-bit Linux. I wonder if the forced conversion to `Int` in `confusion` means that something in `randindex` overflows when `Int` is only half as big.

File: test/randindex.jl [hunk 1]
```diff
@@ -34,4 +34,9 @@ a3 = [3, 3, 3, 2, 2, 2, 1, 1, 1, 1]
 
 @test randindex(a1, a2) == randindex(a2, a1)
 
+@test randindex(ones(Int, 3), ones(Int, 3)) == (1, 1, 0, 1)
+
+a,b = rand(1:5,100_000), rand(1:5,100_000)
```
Comment (pcjentsch) [inline test/randindex.jl:39]: Yes the problem is this test is too big for 32 bit systems. We can just remove a zero from the length of these vectors and it's fine.

```
a,b = rand(1:5,10000), rand(1:5,10000)
```

I'm not sure if it's worth it to add a check in `confusion`.

**Comment by pcjentsch** [comment]
Hm, not sure why that change broke the test. Sorry about that, I did test it locally before suggesting it.

File: src/confusion.jl [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    confusion(a::AbstractVector{<:Integer}, b::AbstractVector{<:Integer}) -> Matrix{Int}
+
+Return 2x2 confusion matrix `C` that represents partition co-occurrence or
+similarity matrix between two clusterings by considering all pairs of samples
+and counting  pairs that are assigned into the same or into different clusters
+under the true and predicted clusterings.
+
+Considering a pair of samples that is in the same group as a **positive pair**,
+and a pair is in the different group as a **negative pair**, then the count of
+true positives is `C₀₀`, false negatives is `C₀₁`, false positives `C₁₀`, and
```
Comment (alyst) [inline src/confusion.jl:14]: Could you please update it to 1-based indexing of the matrix `C`?

File: src/randindex.jl [hunk 3]
```diff
@@ -18,35 +18,25 @@ Returns a tuple of indices:
...
-    nis = sum(abs2, sum(c, dims=2))        # sum of squares of sums of rows
-    njs = sum(abs2, sum(c, dims=1))        # sum of squares of sums of columns
-
-    t1 = binomial(n, 2)                    # total number of pairs of entities
-    t2 = sum(abs2, c)                      # sum over rows & columnns of nij^2
-    t3 = .5*(nis+njs)
-
-    # Expected index (for adjustment)
-    nc = (n*(n^2+1)-(n+1)*nis-(n+1)*njs+2*(nis*njs)/n)/(2*(n-1))
+    a, c, b, d = confusion(a,b) # Table 2 from Steinley 2004
```
Comment (alyst) [inline src/randindex.jl:26]:
```suggestion
    c11, c21, c12, c22 = confusion(a, b) # Table 2 from Steinley 2004
```

File: src/randindex.jl [hunk 4]
```diff
@@ -18,35 +18,25 @@ Returns a tuple of indices:
...
 
-    A = t1+t2-t3;        # agreements count
-    D = -t2+t3;          # disagreements count
+    t = a+ b + c + d   # total number of pairs of entities
+    A = a + d
+    D = b + c
```
Comment (alyst) [inline src/randindex.jl:30]:
```suggestion
    t = c12 + c21 + c12 + c22   # total number of pairs of entities
    A = c11 + c22
    D = c12 + c21
```

File: test/randindex.jl [hunk 5]
```diff
@@ -34,4 +34,9 @@ a3 = [3, 3, 3, 2, 2, 2, 1, 1, 1, 1]
 
 @test randindex(a1, a2) == randindex(a2, a1)
 
+@test randindex(ones(Int, 3), ones(Int, 3)) == (1, 1, 0, 1)
+
+a,b = rand(1:5,10_000), rand(1:5,10_000)
+@test randindex(a,b)[1] < 1.0e-2
```
Comment (alyst) [inline test/randindex.jl:40]:
```suggestion
a, b = rand(1:5, 10_000), rand(1:5, 10_000)
@test randindex(a, b)[1] < 1.0e-2
```

**Review by alyst — COMMENTED** [review]
Thanks for the PR!
Looks good. Could you please implement the suggestions (which are basically just formatting improvements)?

**Comment by alyst** [comment]
@wildart So sorry I overlooked your updates! Thank you!

**Pull request merged**